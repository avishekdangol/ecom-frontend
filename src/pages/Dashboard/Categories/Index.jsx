import { useEffect } from 'react';
import {
  Button, Input, Modal,
} from 'antd';
import { Form, Formik } from 'formik';
import jwt from '@/auth/useJwt';
import DashLayout from '@/layouts/DashLayout';
import AntTable from '@/components/reusables/AntTable';
import showNotification from '@/utils/Toasts';
import CategorySchema from './validations/CategorySchema';
import ProcessingSpinButton from '@/components/reusables/ProcessingSpinButton';
import {
  columns,
  fetchCategories,
  getCategoriesList,
  getShowAddCategoryModal,
  toggleAddCategoryModal,
} from './useCategories.js';

function Categories() {
  useEffect(() => { fetchCategories(); }, []);
  return (
    <DashLayout>
      <AntTable
        title="Categories"
        data={getCategoriesList()}
        addData={toggleAddCategoryModal}
        refetchData={fetchCategories}
        columns={columns}
        updateApi="updateCategory"
        deleteApi="deleteCategory"
        bulkDeleteApi="deleteBulkCategories"
      />

      <Modal
        open={getShowAddCategoryModal()}
        title="Add New Category"
        footer={null}
        onCancel={() => { toggleAddCategoryModal(false); }}
      >
        <Formik
          initialValues={{
            name: '',
          }}
          validationSchema={CategorySchema}
          onSubmit={(values, actions) => {
            jwt.storeCategory(values).then((response) => {
              showNotification('success', response);
              fetchCategories();
              toggleAddCategoryModal(false);
              actions.resetForm();
            }).catch(({ response }) => {
              if (response) {
                const errors = Object.values(response.data.errors);
                showNotification('error', response, errors[0]);
              }
            });
          }}
        >
          {({
            values, touched, errors, handleChange, handleSubmit, isSubmitting, resetForm,
          }) => (
            <Form>
              <div className="mb-4">
                <Input
                  name="name"
                  type="text"
                  value={values.name}
                  placeholder="Category Name"
                  onPressEnter={handleSubmit}
                  onChange={handleChange}
                />
                <p className="mb-0">
                  {touched.name && errors.name && <small className="text-red-500">{errors.name}</small>}
                </p>
              </div>

              <div className="flex justify-end">
                <Button
                  disabled={isSubmitting}
                  onClick={() => { toggleAddCategoryModal(false); resetForm(); }}
                >
                  Cancel
                </Button>

                <ProcessingSpinButton
                  buttonClasses="primary-btn ml-2"
                  condition={isSubmitting}
                  text="Add"
                  action={handleSubmit}
                />
              </div>
            </Form>
          )}
        </Formik>
      </Modal>
    </DashLayout>
  );
}
export default Categories;
