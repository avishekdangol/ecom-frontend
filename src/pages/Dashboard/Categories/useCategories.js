import { signal } from '@preact/signals-react';
import jwt from '@/auth/useJwt';
import showNotification from '@/utils/Toasts';

const categories = signal([]);
const categoriesMeta = signal({});
const showAddCategoryModal = signal(false);

export const fetchCategories = () => {
  jwt.getCategories().then((response) => {
    const { data, ...rest } = response.data.data;

    categories.value = data;
    categoriesMeta.value = rest;
  }).catch(({ response }) => {
    if (response) {
      const errors = Object.values(response.data.errors);
      showNotification('error', response, errors[0]);
    }
  });
};

export const toggleAddCategoryModal = (show = true) => {
  showAddCategoryModal.value = show;
};

export const getCategoriesList = () => categories.value;
export const getCategoriesMeta = () => categoriesMeta.value;
export const getShowAddCategoryModal = () => showAddCategoryModal.value;

export const columns = [
  {
    title: 'Name',
    dataIndex: 'name',
    width: '25%',
    editable: true,
  },
];
