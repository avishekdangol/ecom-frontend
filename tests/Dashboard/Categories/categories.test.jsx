import {
  render, screen, fireEvent, waitFor,
} from '@root/tests/testUtils';
import Categories from '@/pages/Dashboard/Categories/Index';
import useCategories from '@/pages/Dashboard/Categories/useCategories';

jest.mock('@/utils/common', () => ({
  getUserData: {
    value: {
      role: 'admin',
    },
  },
}));

jest.mock('@/pages/Dashboard/Categories/useCategories.js', () => ({
  columns: [{
    title: 'Name',
    dataIndex: 'name',
    width: '25%',
    editable: true,
  }],
  fetchCategories: jest.fn(),
  getCategoriesList: jest.fn(() => []),
  getShowAddCategoryModal: jest.fn(() => false),
  toggleAddCategoryModal: jest.fn(),
}));

describe('Categories', () => {
  test('snapshot', () => {
    const { asFragment } = render(<Categories />);
    expect(asFragment()).toMatchSnapshot();
  });

  test('category page loads successfully', () => {
    render(<Categories />);
    const categoriesTitle = screen.getByRole('heading', { level: 1, name: /categories/i });

    expect(categoriesTitle).toBeInTheDocument();
  });

  test('category page has add category button', () => {
    render(<Categories />);
    const addCategoryBtn = screen.getByRole('button', { name: /add categories/i });

    expect(addCategoryBtn).toBeInTheDocument();
  });

  // test('add category modal opens when the add category button is clicked', async () => {
  //   render(<Categories />);
  //   const addCategoryBtn = screen.getByRole('button', { name: /add categories/i });

  //   fireEvent.click(addCategoryBtn);

  //   const addNewCategoryTitle = await waitFor(() => screen.getByText(/add new/i));

  //   expect(addNewCategoryTitle).toBeInTheDocument();
  // });

  test('show no data message when the table is empty', () => {
    render(<Categories />);
    const emptyMessage = screen.getByText(/no data/i);

    expect(emptyMessage).toBeInTheDocument();
  });

  test('show data when the table has data', () => {
    useCategories.getCategoriesList.mockImplementation(() => [{ id: 1, name: 'Trainers' }]);
    render(<Categories />);
    const data = screen.getByText(/trainers/i);

    expect(data).toBeInTheDocument();
  });
});
