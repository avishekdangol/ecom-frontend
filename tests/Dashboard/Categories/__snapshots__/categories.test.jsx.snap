// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Categories snapshot 1`] = `
<DocumentFragment>
  <div
    class="ant-space css-dev-only-do-not-override-1lp5z9r ant-space-vertical ant-space-gap-row-small ant-space-gap-col-small bg-default"
    style="width: 100%;"
  >
    <div
      class="ant-space-item"
    >
      <div
        class="ant-layout css-dev-only-do-not-override-1lp5z9r"
      >
        <header
          style="background-color: rgb(16, 41, 37); color: rgb(230, 230, 230); padding: 10px 50px;"
        >
          <div
            class="ant-row justify-between items-center css-dev-only-do-not-override-1lp5z9r"
          >
            <div
              class="ant-col css-dev-only-do-not-override-1lp5z9r"
            >
              <a
                class="flex items-center"
                href="/"
              >
                <svg
                  baseProfile="tiny"
                  class="text-2xl mb-1"
                  fill="currentColor"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  version="1.2"
                  viewBox="0 0 24 24"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M12 3s-6.186 5.34-9.643 8.232c-.203.184-.357.452-.357.768 0 .553.447 1 1 1h2v7c0 .553.447 1 1 1h3c.553 0 1-.448 1-1v-4h4v4c0 .552.447 1 1 1h3c.553 0 1-.447 1-1v-7h2c.553 0 1-.447 1-1 0-.316-.154-.584-.383-.768-3.433-2.892-9.617-8.232-9.617-8.232z"
                  />
                </svg>
                 
                <h3
                  class="ml-1"
                >
                  Pahichaan
                </h3>
              </a>
            </div>
            <div
              class="ant-col css-dev-only-do-not-override-1lp5z9r"
            >
              <div
                class="flex items-center"
              >
                <div
                  class="text-right"
                >
                  <h3 />
                  <p
                    class="capitalize mb-0"
                  >
                    admin
                  </p>
                </div>
                <div
                  class="ant-space css-dev-only-do-not-override-1lp5z9r ant-space-horizontal ant-space-align-center ant-space-gap-row-small ant-space-gap-col-small ant-dropdown-trigger"
                >
                  <div
                    class="ant-space-item"
                  >
                    <span
                      class="ant-avatar ant-avatar-circle cursor-pointer bg-teal-700 flex items-center ml-4 css-dev-only-do-not-override-1lp5z9r"
                      data-testid="user-menu-btn"
                    >
                      <span
                        class="ant-avatar-string"
                        style="transform: scale(1) translateX(-50%);"
                      >
                        <svg
                          fill="currentColor"
                          height="20"
                          stroke="currentColor"
                          stroke-width="0"
                          viewBox="0 0 24 24"
                          width="20"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            d="M7.5 6.5C7.5 8.981 9.519 11 12 11s4.5-2.019 4.5-4.5S14.481 2 12 2 7.5 4.019 7.5 6.5zM20 21h1v-1c0-3.859-3.141-7-7-7h-4c-3.86 0-7 3.141-7 7v1h17z"
                          />
                        </svg>
                      </span>
                    </span>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </header>
        <div
          class="ant-row css-dev-only-do-not-override-1lp5z9r"
        >
          <div
            class="ant-col ant-col-3 css-dev-only-do-not-override-1lp5z9r"
            style="background-color: rgb(17, 44, 40); min-height: 92.2vh;"
          >
            <div
              class="dashboard-sidebar"
            >
              <a
                href="/admin/dashboard"
              >
                <div
                  class="sidebar-item flex items-center p-4 "
                >
                  <div
                    class="text-xl mb-1"
                  >
                    <svg
                      fill="currentColor"
                      height="1em"
                      stroke="currentColor"
                      stroke-width="0"
                      viewBox="0 0 24 24"
                      width="1em"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M0 0h24v24H0z"
                        fill="none"
                      />
                      <path
                        d="M3 13h8V3H3v10zm0 8h8v-6H3v6zm10 0h8V11h-8v10zm0-18v6h8V3h-8z"
                      />
                    </svg>
                  </div>
                  <h6
                    class="ml-1"
                  >
                    Dashboard
                  </h6>
                </div>
              </a>
              <a
                href="/admin/categories"
              >
                <div
                  class="sidebar-item flex items-center p-4 "
                >
                  <div
                    class="text-xl mb-1"
                  >
                    <svg
                      fill="currentColor"
                      height="1em"
                      stroke="currentColor"
                      stroke-width="0"
                      viewBox="0 0 24 24"
                      width="1em"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M0 0h24v24H0z"
                        fill="none"
                      />
                      <path
                        d="M12 2l-5.5 9h11z"
                      />
                      <circle
                        cx="17.5"
                        cy="17.5"
                        r="4.5"
                      />
                      <path
                        d="M3 13.5h8v8H3z"
                      />
                    </svg>
                  </div>
                  <h6
                    class="ml-1"
                  >
                    Categories
                  </h6>
                </div>
              </a>
              <a
                href="/admin/products"
              >
                <div
                  class="sidebar-item flex items-center p-4 "
                >
                  <div
                    class="text-xl mb-1"
                  >
                    <svg
                      fill="currentColor"
                      height="1em"
                      stroke="currentColor"
                      stroke-width="0"
                      viewBox="0 0 512 512"
                      width="1em"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M32 32H480c17.7 0 32 14.3 32 32V96c0 17.7-14.3 32-32 32H32C14.3 128 0 113.7 0 96V64C0 46.3 14.3 32 32 32zm0 128H480V416c0 35.3-28.7 64-64 64H96c-35.3 0-64-28.7-64-64V160zm128 80c0 8.8 7.2 16 16 16H336c8.8 0 16-7.2 16-16s-7.2-16-16-16H176c-8.8 0-16 7.2-16 16z"
                      />
                    </svg>
                  </div>
                  <h6
                    class="ml-1"
                  >
                    Products
                  </h6>
                </div>
              </a>
            </div>
          </div>
          <div
            class="ant-col ant-col-21 css-dev-only-do-not-override-1lp5z9r"
          >
            <main
              class="ant-layout-content css-dev-only-do-not-override-1lp5z9r"
              style="min-height: 70vh; color: rgb(0, 0, 0);"
            >
              <section
                class="p-4"
              >
                <div
                  class="flex justify-between items-center w-full"
                >
                  <h1>
                    Categories
                  </h1>
                  <div
                    class="flex"
                  >
                    <button
                      class="ant-btn css-dev-only-do-not-override-1lp5z9r ant-btn-default mx-2"
                      type="button"
                    >
                      <span>
                        Add Categories
                      </span>
                    </button>
                  </div>
                </div>
              </section>
              <div
                class="ant-table-wrapper css-dev-only-do-not-override-1lp5z9r"
              >
                <div
                  class="ant-spin-nested-loading css-dev-only-do-not-override-1lp5z9r"
                >
                  <div
                    class="ant-spin-container"
                  >
                    <div
                      class="ant-table ant-table-bordered ant-table-empty"
                    >
                      <div
                        class="ant-table-container"
                      >
                        <div
                          class="ant-table-content"
                        >
                          <table
                            style="table-layout: auto;"
                          >
                            <colgroup>
                              <col
                                class="ant-table-selection-col"
                              />
                              <col
                                style="width: 25%;"
                              />
                            </colgroup>
                            <thead
                              class="ant-table-thead"
                            >
                              <tr>
                                <th
                                  class="ant-table-cell ant-table-selection-column"
                                  scope="col"
                                >
                                  <div
                                    class="ant-table-selection"
                                  >
                                    <label
                                      class="ant-checkbox-wrapper ant-checkbox-wrapper-disabled css-dev-only-do-not-override-1lp5z9r"
                                    >
                                      <span
                                        class="ant-checkbox ant-wave-target css-dev-only-do-not-override-1lp5z9r ant-checkbox-disabled"
                                      >
                                        <input
                                          aria-label="Select all"
                                          class="ant-checkbox-input"
                                          disabled=""
                                          type="checkbox"
                                        />
                                        <span
                                          class="ant-checkbox-inner"
                                        />
                                      </span>
                                    </label>
                                  </div>
                                </th>
                                <th
                                  class="ant-table-cell"
                                  scope="col"
                                >
                                  Name
                                </th>
                                <th
                                  class="ant-table-cell"
                                  scope="col"
                                >
                                  Operation
                                </th>
                              </tr>
                            </thead>
                            <tbody
                              class="ant-table-tbody"
                            >
                              <tr
                                class="ant-table-placeholder"
                              >
                                <td
                                  class="ant-table-cell"
                                  colspan="3"
                                >
                                  <div
                                    class="css-dev-only-do-not-override-1lp5z9r ant-empty ant-empty-normal"
                                  >
                                    <div
                                      class="ant-empty-image"
                                    >
                                      <svg
                                        height="41"
                                        viewBox="0 0 64 41"
                                        width="64"
                                        xmlns="http://www.w3.org/2000/svg"
                                      >
                                        <g
                                          fill="none"
                                          fill-rule="evenodd"
                                          transform="translate(0 1)"
                                        >
                                          <ellipse
                                            cx="32"
                                            cy="33"
                                            fill="#f5f5f5"
                                            rx="32"
                                            ry="7"
                                          />
                                          <g
                                            fill-rule="nonzero"
                                            stroke="#d9d9d9"
                                          >
                                            <path
                                              d="M55 12.76L44.854 1.258C44.367.474 43.656 0 42.907 0H21.093c-.749 0-1.46.474-1.947 1.257L9 12.761V22h46v-9.24z"
                                            />
                                            <path
                                              d="M41.613 15.931c0-1.605.994-2.93 2.227-2.931H55v18.137C55 33.26 53.68 35 52.05 35h-40.1C10.32 35 9 33.259 9 31.137V13h11.16c1.233 0 2.227 1.323 2.227 2.928v.022c0 1.605 1.005 2.901 2.237 2.901h14.752c1.232 0 2.237-1.308 2.237-2.913v-.007z"
                                              fill="#fafafa"
                                            />
                                          </g>
                                        </g>
                                      </svg>
                                    </div>
                                    <div
                                      class="ant-empty-description"
                                    >
                                      No data
                                    </div>
                                  </div>
                                </td>
                              </tr>
                            </tbody>
                          </table>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </main>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
